### ISSUE TITLE
"No router instance found." error when setting Router.beforePopState in pages/_app.js

### ISSUE THREAD

**Issue by peacefullybound** [zeit/next.js#4289]
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Expected Behavior

Router.beforePopState should be easy to set. The code should reside in pages/_app.js with client-side only guard since it should be run on client only once.

## Current Behavior

I put up an example repository [here](https://gitlab.com/peacefullybound/next-js-custom-routing-test).

Server-side went ok without an error, but browsers complain with the following.

```
No router instance found.
You should only use "next/router" inside the client side of your app.

Error: No router instance found.
You should only use "next/router" inside the client side of your app.

    at throwIfNoRouter (http://localhost:3000/_next/static/commons/main.js:7835:11)
    at Object.SingletonRouter.(anonymous function) [as beforePopState] (http://localhost:3000/_next/static/commons/main.js:7799:5)
    at Object.<anonymous> (http://localhost:3000/_next/-/page/_app.js:2819:55)
    at Object../pages/_app.js (http://localhost:3000/_next/-/page/_app.js:2955:30)
    at __webpack_require__ (http://localhost:3000/_next/static/commons/manifest.js:715:31)
    at fn (http://localhost:3000/_next/static/commons/manifest.js:118:20)
    at Object.1 (http://localhost:3000/_next/-/page/_app.js:2962:18)
    at __webpack_require__ (http://localhost:3000/_next/static/commons/manifest.js:715:31)
    at webpackJsonpCallback (http://localhost:3000/_next/static/commons/manifest.js:26:23)
    at http://localhost:3000/_next/-/page/_app.js:5:7
```

## Steps to Reproduce (for bugs)

[Example repository](https://gitlab.com/peacefullybound/next-js-custom-routing-test)

## Context

See "Expected Behavior" above.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    | 6.0.0 |
| node    | 10.0.0 |
| OS      | macOS High Sierra 10.13.4 |
| browser | Google Chrome 66.0.3359.139 |
| etc     |         |

**Comment by peacefullybound** [comment]
Regardless of my expectation, I want to know if I miss something, or more proper way of setting Router.beforePopState. Thanks in advance.

**Comment by peacefullybound** [comment]
I fixed the issue by moving the usage of Router.beforePopState into App#componentDidMount. I'll close this. Thanks for any concerns.

**Comment by afganhalbana** [comment]
hi @peacefullybound i did move it to componentDidMount lifecycle but unfortunately it still didnt work whenever i changed the route, any thought tho?

**Comment by buiduccuong30051989** [comment]
hihi

**Comment by 2JN** [comment]
I solve my problem by checking first if Router wasn't undefined.
`const pathname = Router ? Router.pathname : ''`

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
